include/libtorrent/tracker_manager.hpp

```diff
@@ -400,7 +400,8 @@ namespace libtorrent
 		mutable mutex m_mutex;
 
 		// maps transactionid to the udp_tracker_connection
-		// TODO: this should be unique_ptr in the future
+		// These must use shared_ptr to avoid a dangling reference
+		// if a connection is erased while a timeout event is in the queue
```

Comment:
yeah, working on the listen socket patch I realized that this pattern isn't that great. It may seem nice to begin with, but I believe components may be more composable if async operations are issued and handled from outside the object itself (like all the built-in asio types). That way the user of a class can control its life-time, including destroying it before all outstanding operations have come back